types/tx.go

```diff
@@ -193,39 +193,71 @@ func ComputeProtoSizeForTxs(txs []Tx) int64 {
 	return int64(pdData.Size())
 }
 
-// UnwrapMalleatedTx attempts to unmarshal the provided transaction into a malleated
-// transaction wrapper, if this can be done, then it returns true. A malleated
-// transaction is a normal transaction that has been derived (malleated) from a
-// different original transaction. The returned hash is that of the original
-// transaction, which allows us to remove the original transaction from the
-// mempool. NOTE: protobuf sometimes does not throw an error if the transaction
-// passed is not a tmproto.MalleatedTx, since the schema for PayForMessage is kept
-// in the app, we cannot perform further checks without creating an import
-// cycle.
+// UnwrapMalleatedTx attempts to unmarshal the provided transaction into a
+// malleated transaction. It returns true if the provided transaction is a
+// malleated transaction. A malleated transaction is a transaction that contains
+// a MsgPayForBlob that has been wrapped with a share index.
+//
+// NOTE: protobuf sometimes does not throw an error if the transaction passed is
+// not a tmproto.MalleatedTx, since the schema for PayForBlob is kept in the
+// app, we cannot perform further checks without creating an import cycle.
 func UnwrapMalleatedTx(tx Tx) (malleatedTx tmproto.MalleatedTx, isMalleated bool) {
 	// attempt to unmarshal into a a malleated transaction
 	err := proto.Unmarshal(tx, &malleatedTx)
 	if err != nil {
 		return malleatedTx, false
 	}
-	// this check will fail to catch unwanted types should those unmarshalled
-	// types happen to have a hash sized slice of bytes in the same field number
-	// as originalTxHash. TODO(evan): either fix this, or better yet use a different
-	// mechanism
-	if len(malleatedTx.OriginalTxHash) != tmhash.Size {
+	if malleatedTx.TypeId != consts.ProtoMalleatedTxTypeID {
 		return malleatedTx, false
 	}
 	return malleatedTx, true
 }
 
-// WrapMalleatedTx creates a wrapped Tx that includes the original transaction's hash
-// so that it can be easily removed from the mempool. note: must be unwrapped to
-// be a viable sdk.Tx
-func WrapMalleatedTx(originalHash []byte, shareIndex uint32, malleated Tx) (Tx, error) {
+// WrapMalleatedTx creates a wrapped Tx that includes the original transaction
+// and the share index of the start of its blob.
+//
+// NOTE: must be unwrapped to be a viable sdk.Tx
+func WrapMalleatedTx(shareIndex uint32, tx Tx) (Tx, error) {
 	wTx := tmproto.MalleatedTx{
-		OriginalTxHash: originalHash,
-		Tx:             malleated,
-		ShareIndex:     shareIndex,
+		Tx:         tx,
+		ShareIndex: shareIndex,
+		TypeId:     consts.ProtoMalleatedTxTypeID,
 	}
 	return proto.Marshal(&wTx)
 }
+
+// UnwrapBlobTx attempts to unmarshal a transaction into blob transaction. If an
+// error is thrown, false is returned.
+func UnwrapBlobTx(tx Tx) (bTx tmproto.BlobTx, isBlob bool) {
```

Comment:
I was confused on https://github.com/celestiaorg/celestia-core/pull/893/files#r1029054651 because I interpreted this function as the reverse operation of `WrapBlobTx`. In other words, given a BlobTx return the constituent parts (the blobs and the wrapped Tx). But on closer inspection, this function just unmarshals a BlobTx so proposal to rename? 

```suggestion
// UnmarshalBlobTx attempts to unmarshal a transaction into a BlobTx. If an
// error is thrown, false is returned.
func UnmarshalBlobTx(tx Tx) (bTx tmproto.BlobTx, isBlob bool) {
```

Reply:
good idea https://github.com/celestiaorg/celestia-core/pull/893/commits/263ea4620700240f678ad7c56bc99d882da32faa